src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,166 @@
+            head.next = tail;
+        } else {
+            tail.next = new Node<>(tail, value, null);
+            tail = tail.next;
+        }
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (size == index) {
+            add(value);
+            return;
+        }
+        Node<T> node = getNodeByIndex(index);
+        Node<T> newNode = new Node<>(node.prev, value, node);
+        if (index == 0) {
+            head.prev = newNode;
+            head = newNode;
+            size++;
+            return;
+        }
+        newNode.prev.next = newNode;
+        node.prev = newNode;
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T element: list) {
+            tail.next = new Node<>(tail, element, null);
+            tail = tail.next;
+        }
+        size += list.size();
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T oldItem;
+        if (isIndexValid(index)) {
+            Node<T> editNode = getNodeByIndex(index);
+            oldItem = editNode.item;
+            editNode.item = value;
+        } else {
+            throw new IndexOutOfBoundsException();
+        }
+        return oldItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T value = get(index);
+        if (index == 0 && size == 1) {
+            head = null;
+            tail = null;
+            size = 0;
+            return value;
+        }
+        Node<T> node = getNodeByIndex(index);
+        if (index == 0) {
+            head = null;
+            head = node.next;
+        } else if (index == size - 1) {
+            tail = null;
+            tail = node.prev;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
+        size--;
+        return value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        int index = getItemIndex(object);
+        if (index == -1) {
+            return false;
+        }
+        remove(index);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        T item;
+        Node<T> next;
+        Node<T> prev;
+
+        Node(Node<T> prev, T item, Node<T> next) {
+            this.prev = prev;
+            this.item = item;
+            this.next = next;
+        }
+    }
+
+    private boolean isIndexValid(int index) {
+        return index >= 0 && index < size;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        if (!isIndexValid(index)) {
```

Comment:
better let's create method `checkIndexIsValid(index)` that will throw an exception if the index is not valid